tracer/test/Datadog.Trace.DuckTyping.Tests/DuckChainingNullableTests.cs

```diff
@@ -30,6 +32,33 @@ public void CanDuckChainNullableDirectly()
             var proxy = DuckType.CreateCache<NullableProxyTestStruct>.CreateFrom(instance);
         }
 
+        [Fact]
+        public void CanDuckChainWithNullInstance()
+        {
+            var firstClass = new FirstClass();
+            var firstClassProxy = firstClass.DuckCast<IFirstClass>();
+
+            firstClass.Value.Should().BeNull();
```

Comment:
Ok, I think this actually worries me... Up to now, we've always assumed that the proxy is always non-null. To check if the _instance_ is null, you need to type as IDuckType and check if the `Instance` property is null.

We have a lot of assumptions in our integrations that the proxy type _will never be null_. This seems to be changing that assumption? 😟 Seems like a potential recipe for null refs 😬 

Unless I'm confused. I'm often confused about this stuff

Reply:
Yup confirmed, running the same test on `master` passes.

Reply:
> The DuckChaining with an interface can return null

I think I didn't realise that, and thought it would always be non-null 😬 I forgot the duck-chaining works differently... cc @kr-igor because I may have lied to you 🙈 

> The DuckChaining with an DuckCopy struct cannot return null. That's why I'm introducing the support to Nullable<T> in duckchaining allowing us to use the DuckCopy struct with the nullable operator ?

This behaviour difference seems weird to me. Why do we have it that the struct can't return null, but the interface can? In both case we create a proxy struct right? IMO it would be better to make the behaviour consistent between them (i.e. never return null) rather than add another confusion point (support for nullables), though we could certainly do both - at least making it nullable is easy to understand.

> I believe the behavior that you are describing is the one for the ducktyping using the generic constraints in the calltarget callbacks. That one is never going to be null, because we create the generic argument with the proxy type that is a struct.

Yeah, I thought we _always_ did this

Reply:
>This behaviour difference seems weird to me. Why do we have it that the struct can't return null, but the interface can?
Reasons :) 

1. Ducktype tries to reduce heap allocations always. 
2. This is a consequence of returning `default` when doing the null check of the original value. ValueType will have the empty default value and classes will have null.
3. Returning always an instance is not always possible. DuckType also accepts [Abstract and Virtual classes](https://github.com/DataDog/dd-trace-dotnet/blob/master/docs/development/DuckTyping.md#proxy-types) in addition to structs.

>In both case we create a proxy struct right? 
Not really, in some cases before creating the struct proxy type we need to check the actual instance value. If this actual instance value is null we cannot create the proxy.

This is done because the the return type of an original type could be an abstract or base class or an interface, and we want to do the proxy to the actual instance type. Due to some of the data we want to access may be on the instance type and not in the declared type.

See: https://github.com/DataDog/dd-trace-dotnet/blob/master/tracer/src/Datadog.Trace/DuckTyping/DuckType.cs#L1300-L1315

>IMO it would be better to make the behaviour consistent between them (i.e. never return null) rather than add another confusion point (support for nullables), though we could certainly do both - at least making it nullable is easy to understand.

The problem is that by not having the proxy type we cannot create an instance of something that implements your `ISomething`...

I don't know, if you remove the `nullable` checks feels natural that a proxy with a class type on their properties could return null, and a proxy with a duck copy struct in their return type because is a struct will always have some value, and after this change if you want to keep the same behavior just add  `?` to your duck copy struct declaration.

Reply:
> DuckType also accepts [Abstract and Virtual classes](https://github.com/DataDog/dd-trace-dotnet/blob/master/docs/development/DuckTyping.md#proxy-types) in addition to structs.

If it introduces this ambiguity, and we don't recommend it for perf reasons, maybe it shouldn't? 😄 

> Due to some of the data we want to access may be on the instance type and not in the declared type.

Good point 🤔 

> If you remove the nullable checks feels natural that a proxy with a class type on their properties could return null, and a proxy with a duck copy struct in their return type because is a struct will always have some value, and after this change if you want to keep the same behavior just add ? to your duck copy struct declaration.

I guess I disagree because I thought if we had a ducktype interface, _we create a struct behind the scenes_. 

IMO it makes everything much more confusing knowing _when_ a duck type _might_ be `null`, and when it will _never_ be null, and instead you will always need to check `IDuckType.Instance`... 

I think we should probably document all these cases in the ducktyping docs either way, because I've definitely been giving incorrect advice 😬

Reply:
I see, the thing is we cannot create an instance of the proxy if we don't have proxy yet, so for classes I don't see how can we do that. Also, remember `IDuckType.Instance` cannot be used when we have duckchaining using `[DuckCopy]` struct. So, there's not a single way of doing the same thing in both scenarios.

Also one thing I like of supporting `Nullable<T>` where T is [DuckCopy] struct is in this scenario:

`FirstProxy.SecondProxyStruct.Name` if thi results `null` we don't know what's null in the original object, it could be `Name` or `Second` due the fact `SecondProxyStruct` will always be initialized.

Now and just remember that is `optional`. you can add the nullability operator to `SecondProxyStruct` and now I can check if `FirstProxy.SecondProxyStruct` is null meaning that `Second` in the original object is null. It just gives you more information. As I said that's optional, the current behavior will continued be respected if you don't do anything.

I agree we have to document this very well and keep in mind that `DuckCopy` structs are a totally different breed (DuckChaining behavior, Fields instead properties, etc).

Reply:
> Also, remember IDuckType.Instance cannot be used when we have duckchaining using [DuckCopy] struct.

Oh really?! I _didn't_ know that 😅 That changes all sorts of things, for example: 

> FirstProxy.SecondProxyStruct.Name if thi results null we don't know what's null in the original object

I thought we were able to go `((IDuckType)FirstProxy.SecondProxyStruct).Instance`, but I guess not then? 😅 

> I agree we have to document this very well
➕➕➕
This is probably the biggest thing tbh, whatever we do, I think we desperately need very detailed docs (and examples) for this. The tests are great, but not easy to refer to 🙂